### PR TITLE
REGR: ensure numpy fixed-size binary/string arrays are converted to object dtype for Index creation

### LINKED ISSUE #57645: BUG: Cannot use numpy FLS as indicies since pandas 2.2.1
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [X] I have confirmed this bug exists on the [main branch](https://pandas.pydata.org/docs/dev/getting_started/install.html#installing-the-development-version-of-pandas) of pandas.

### Reproducible Example

```python
import numpy as np
import pandas as pd

# List of fixed-length strings
fixed_strings = ["apple", "banana", "orange", "grape"]

# Define the fixed length for the strings
string_length = 6  # Adjust as needed

# Create NumPy array of fixed-length strings
arr = np.array(fixed_strings, dtype=f"S{string_length}")

df = pd.DataFrame(pd.Series(arr), columns=["fruit"])

# Raises NotImplementedError: |S6
df.set_index("fruit", inplace=True)
```

### Issue Description

The issue shown in the provided code is that when attempting to set the index of the Pandas DataFrame using `df.set_index("fruit", inplace=True)`, a `NotImplementedError` is raised with the message `|S6`.

The reason for this error is that the dtype `|S6` is not supported as an index type in Pandas. When you create a NumPy array with a dtype of fixed-length strings using np.array() and attempt to set it as an index in a DataFrame, Pandas tries to convert it to a suitable index type, but `|S6` is not recognized as a valid option for the index.

This is applicable to all `S` dtypes. The offending logic was introduced in pandas 2.2.1 by these lines of this commit: https://github.com/pandas-dev/pandas/commit/b6fb90574631c84f19f2dbdc68c26d6ce97446b4#diff-fb8a9c322624b0777f3ff7e3ef8320d746b15a2a0b80b7cab3dfbe2e12e06daaR239-R240

The code works as expected in pandas 2.2.0.

### Expected Behavior

What happens when using 2.2.0:

```
import numpy as np
import pandas as pd

# List of fixed-length strings
fixed_strings = ["apple", "banana", "orange", "grape"]

# Define the fixed length for the strings
string_length = 6  # Adjust as needed

# Create NumPy array of fixed-length strings
arr = np.array(fixed_strings, dtype=f"S{string_length}")

df = pd.DataFrame(pd.Series(arr), columns=["fruit"])

df.set_index("fruit", inplace=True)

print(df)
```

Prints:

```
Empty DataFrame
Columns: []
Index: [b'apple', b'banana', b'orange', b'grape']
```

### Installed Versions

<details>

INSTALLED VERSIONS
------------------
commit                : bdc79c146c2e32f2cab629be240f01658cfb6cc2
python                : 3.11.1.final.0
python-bits           : 64
OS                    : Linux
OS-release            : 6.6.10-76060610-generic
Version               : #202401051437~1704728131~22.04~24d69e2 SMP PREEMPT_DYNAMIC Mon J
machine               : x86_64
processor             : x86_64
byteorder             : little
LC_ALL                : None
LANG                  : en_GB.UTF-8
LOCALE                : en_GB.UTF-8

pandas                : 2.2.1
numpy                 : 1.26.4
pytz                  : 2024.1
dateutil              : 2.8.2
setuptools            : 67.7.2
pip                   : 24.0
Cython                : None
pytest                : 7.4.4
hypothesis            : None
sphinx                : None
blosc                 : None
feather               : 0.4.1
xlsxwriter            : None
lxml.etree            : 4.9.3
html5lib              : None
pymysql               : None
psycopg2              : None
jinja2                : 3.1.2
IPython               : 8.15.0
pandas_datareader     : None
adbc-driver-postgresql: None
adbc-driver-sqlite    : None
bs4                   : 4.12.2
bottleneck            : None
dataframe-api-compat  : None
fastparquet           : None
fsspec                : 2023.10.0
gcsfs                 : None
matplotlib            : 3.8.3
numba                 : 0.59.0
numexpr               : 2.9.0
odfpy                 : None
openpyxl              : 3.1.2
pandas_gbq            : None
pyarrow               : 15.0.0
pyreadstat            : None
python-calamine       : None
pyxlsb                : None
s3fs                  : 2023.10.0
scipy                 : 1.12.0
sqlalchemy            : 2.0.20
tables                : 3.9.2
tabulate              : 0.9.0
xarray                : None
xlrd                  : None
zstandard             : None
tzdata                : 2023.3
qtpy                  : None
pyqt5                 : None
None

</details>

### PR DESCRIPTION
**Pull request by PF2100**
- [X] closes #57645 
- [X] [Tests added and passed] if fixing a bug or adding a new feature
- [X] All [code checks passed].
- [ ] Added [type annotations] to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

While using the function set_index with parameter inplace=True, the function would try and create a new index where its dtype would be a FLS `S{value}` dtype, which was not recognized by the function `_dtype_to_subclass` and raised a NotImplementedError. That said , by adding a verification that recognizes FLS dtype , the index is created successfully and the function executes properly.

### CODE REVIEW

File: pandas/tests/frame/methods/test_set_index.py [hunk 1]
```diff
@@ -617,6 +617,20 @@ def test_set_index_raise_on_len(
         with pytest.raises(ValueError, match=msg):
             df.set_index(["A", df.A, box(values)], drop=drop, append=append)
 
+    def test_set_index_with_FLS_Dtype(self):
+        string_length = 6
+        in_dtype, df_name = f"S{string_length}", "fruit"
+        data = ["apple", "banana", "orange", "grape"]
+
+        # Create array with FLS(|S{value}) dtype
+        arr = np.array(data, dtype=in_dtype)
```
Comment (jorisvandenbossche) [inline pandas/tests/frame/methods/test_set_index.py:626]: For the purpose of the tests, we can simplify the snippet from the issue report a bit
```suggestion
        # Create array with FLS(|S{value}) dtype
        arr = np.array( ["apple", "banana", "orange", "grape"], dtype="S6")
```

**Review by jorisvandenbossche — COMMENTED** [review]
Thanks a lot for working on this, and apologies for the lack of reviews!

The original PR that introduced this change edited the `array_tuplesafe` function (https://github.com/pandas-dev/pandas/pull/57139), which is called inside `Index.__new__` before calling `_dtype_to_subclass`. So if we would change `array_tuplesafe` to still return object dtype instead of numpy S/U, that would fix it before getting to `_dtype_to_subclass`. 

The current fix also essentially allows to create an Index with an underlying numpy array with S6 dtype, while I think we want to restore the behaviour of converting to object dtype (I think my suggestion above should do that)


Other comment: could you also add a test for direct Index constructor from a Series object (i.e. `pd.Index(pd.Series(arr))`), because that is affected as well, in addition to `set_index`

File: pandas/tests/frame/methods/test_set_index.py (hunk 1 above)
Comment (PF2100) [reply pandas/tests/frame/methods/test_set_index.py:626]: Thank you very much for the sugestion and I will change the test. Regarding the asarray_tuplesafe, I went into a little more detail in this PR's issue (#57645) but looking back into the original PR, the line `elif isinstance(values, ABCSeries)` was added which allows for any Series to keep their original types and values without performing any check if their type is compatible or not.

Could a possible solution be something of the sort:

```python
elif isinstance(values, ABCSeries) and values.dtype.kind != "S":
        return values_values
``` 

By refactoring this line, the elif will fail if the values have dtype ="S" and creates a new array with dtype object as it used to. Before I proceed, could I know your thoughts on the suggestion I've made?

**Comment by PF2100** [comment]
@jorisvandenbossche Hi, sorry to bother you, but would it be possible to get some feedback on this? 
Thank you!

**Pull request closed without merging**